man/man3/h2timerLib.3
.\" Copyright (c) 1999-2005 CNRS/LAAS
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd January 21, 2005
.Dt h2timerLib 3
.Os pocoLibs
.Sh NAME 
.Nm h2timerLib
.Nd pocoLibs timers manipulation library
.Sh SYNOPSIS
.Fd "#include <portLib.h>"
.Fd "#include <h2timerLib.h>"
.Ft STATUS
.Fn h2timerInit void
.Ft STATUS
.Fn h2timerEnd void
.Ft H2TIMER_ID
.Fn h2timerAlloc void
.Ft STATUS
.Fn h2timerStart "H2TIMER_ID timerId" "int period" "int delay"
.Ft STATUS
.Fn h2timerPause "H2TIMER_ID timerId"
.Ft STATUS
.Fn h2timerPauseReset "H2TIMER_ID timerId"
.Ft STATUS
.Fn h2timerStop "H2TIMER_ID timerId"
.Ft STATUS
.Fn h2timerChangePeriod "H2TIMER_ID timerId" "int period"
.Ft STATUS
.Fn h2timerFree "H2TIMER_ID timerId"
.Sh DESCRIPTION
.Fn h2timerInit
initializes the timer component of 
.Nm pocolibs. This function is normally called by the global
initialization function
.Xr h2initGlob 3 .
.Pp
.Fn h2timerEnd
cleans up the timer library and frees the resources used.
.Pp
.Fn h2timerAlloc
allocates a new inactive timer. 
.Fn h2timerStart
arms a timer that will trigger after 
.Fa delay
ticks and then every
.Fa period 
ticks per second.
.Fa period
must be a multiple or a integer divisor of 
.Fa delay 
for implementation reasons. 
.Pp
.Fn h2timerPause
blocks the current task until the next expiration of the timer
specified  by
.Fa timerId . 
This function will not block if it was not called enough times in the
past to ensure that all expirations of the timer have been waited on.
.Pp
.Fn h2timerPauseReset
first flushes the synchronisation semaphore used by the timer and
waits for the next expiration of the timer. This ensures that the task
will at least yield the processor before returning. 
.Pp
.Fn h2timerStop
stops the given timer. No more events will be generated by this
timer. Calling 
.Fn h2timerPause
on a stopped timer has a undefined behaviour.
.Pp
.Fn h2timerChangePeriod
changes the 
.Fa period
of a running timer.
.Pp
.Fn h2timerFree
frees a timer and returns it to the timer pool of pocoLibs. If it was
running, it is first stopped. 
.Sh RETURN VALUES
.Fn h2timerAlloc
returns the identifier of the newly allocated timer or 
.Dv NULL
in case of an error.
.Pp
.Fn other functions return 
.Dv OK
or
.Dv ERROR
in case of an error and set the task's errno value.
.Sh ERRORS
.Bl -tag -width .Er
.It Bq Er S_h2timerLib_TIMER_NOT_INIT
the timer library has not been initialized. 
.It Bq Er S_h2timerLib_TOO_MUCH_TIMERS
no free timer was available in 
.Fn h2timerAlloc .
.It Bq Er S_h2timerLib_NOT_STOPPED_TIMER
The specified timer to 
.Fn h2timerStart
is already running.
.It Bq Er S_h2timerLib_BAD_PERIOD
The specified period is not a multiple or a divisor of the delay of
the timer. 
.It Bq Er S_h2timerLib_STOPPED_TIMER
The specified timer is not running for 
.Fn h2timerChangePeriod .
.El
.Sh SEE ALSO
.Xr sysLib 3 ,
.Xr taskLib 3 ,
.Xr wdLib 3 .


CHANGELOG
Version 2.8, xx, xxxx, xxxx

104. Nuke RCS ids in source files.
103. Call mlockall() in h2initGlob() when running under Xenomai.
102. Go back to pmap_unset() to unregister pre-existing posterServer service.
101. Fixes for building Xenomai support.
100. Fixes for cross-compilation in OpenEmbedded
 99. Display the ntick field of poster's timestamps in posterShow().
 98. Enforce stack size in taskSpawn() using pthread_attr_setstacksize().

Version 2.7, 16 april 2010

 97. Fix updating 'ntick' field in H2TIME timespamps of posters.
     (fallout from change #94, which broke pom-genom).

Version 2.6, 13 April 2010

 96. On Linux systems, let taskSuspend() throw a backtrace to stderr.
 95. Use a pipe to wait until h2dev lock is ready before launching posterServ.
 94. Implement poster time stamps using H2TIMESPEC instead of H2TIME. 
 93. Don't start posterServ if it cannot find h2 devices.
 92. Add regression tests for h2time conversion functions
 91. Constify posterName arguments in posterLib.
 90. A few more fixes for running modules on Darwin.
 89. Add regression tests for h2 semaphores.
 88. Change h2 semaphores to mutexes instead of counting semaphores.

Version 2.5, 6 April 2009

 87. Update shipped autotools files to the version provided by Fedora 10.
 86. Implement poster access statistics.
 85. posterServ: use svc_unregister() instead of pmap_unset().

Version 2.4.1, 9 September 2008

 84. Fix a stack corruption bug introduced by change 83.

Version 2.4, 11 August 2008

 83. Add a 'taskFromThread(char *name)' function. (Matthieu Gallien).
 83. Assume that target systems have posix timers when cross-compiling
 82. Let autoheader generate pocolibs-config.h.in
 81. Switch repository to git
 80. Fix a race in change #77.
 79. Eliminate more uses of sprintf().
 78. Fix tickLib regression test.

Version 2.3, 31 July 2007

 77. Fix taskDelete hooks handling.
 76. Relax constraint on h2timer periods.
 75. h2 init no longer starts posterServ by default. Use h2 init -p.
 74. Add a flag to h2devInit() to control the launch of posterServ.
     Bump portLib major.
 73. Fix build of posterLib tests programs.
 72. Remove VxWorks/RTAI/LXRT parts.

Version 2.2, 19 December 2006

 71. Let configure find the gnu tar program for make dist.
 70. Remove the xes subsystem, which is no longer used.
 69. Preliminary support for building Xenomai 2.x enabled libs, using the
     POSIX skin.
 68. Remove RTAI/LXRT support in configure scripts
 67. Remove use of X11 events in xes_server.
 66. Remove double clear of h2timer counters (Sylvain Joyeux).
 65. Fix build problems on Darwin.
 64. Provide a correct day of week in H2TIME structure.
 63. Fix h2timeShow() that was broken in multiple ways.

Version 2.1, 11 November 2005

 62. Some basic support for kbuild for RTAI 3.1 kernel modules
     (not working).
 61. Rework error messages handling. No more central h2errorList.
 60. More ANSI C prototypes in header files.
 59. Add a autogen.sh script to generate the configure script and garbage
     collect unneeded files.
 58. Merge missing parts of lxrt support in existing files.
 57. Add missing files for lxrt support.

Version 2.1beta1, 26 August 2005

 56. Use vpath <pattern> instead of VPATH in Makefiles to minimize trouble
     when builddir != srcdir and srcdir has been configured already.
 55. Add a man page for h2time functions.
 54. Add a h2GetTimeSpec() function that returs a timespec-like structure.
     This should be used instead of calling clock_gettime() directly.
 53. Check for clock_gettime in configure scripts
 52. Add regression tests.
 51. Packaging fixes.
 50. Add a pkg-config data file and install it
 49. Write manual pages for portLib
 48. return S_symLib_NOT_IMPLEMENTED for symFindByValue()
 47. sprintf() -> snprintf() conversion
 46. less printf()s in portLib
 45. more comments translation
 44. Makefiles cleanups (Sylvain Joyeux, Matthieu Herrb).
 43. Add a h2semSet() function to force the value of a semaphore (Sylvain
     Joyeux).
 42. Add a posterName() function and a FIO_FRESH ioctl to posterLib (Sylvain
     Joyeux).
 41. Valgrind support (Sylvain Joyeux).
 40. Remove some unused files.
 39. Make headers C++ compatible (Sylvain Joyeux).
 38. Add --{disable,enable}-posix-timer option to configure, to be able
     to force the timer type that is used by tickLib, as the auto-detection
     of useable POSIX timer is not always reliable enough.
 37. Added --with-linux-kernel-includes option to specify the path to
     linux kernel headers when compiling for RTAI.
 36. Revert the change that made the posterId in posterServ an uint64_t.
 35. Autotoolize the build of the poster test programs.
 34. Start over with libtool style shared libraries version numbers.
     Bump all major revisions to make sure nothing bad happens.
 33. Add a configure test for usabilility of POSIX timers for a
     100Hz sysClk.
 32. Add a 'tick' regress test to check the accuracy of sysClk.
     (Linux 2.4 POSIX timers are known to be unable to set a clock
     at 100Hz).
 31. Some cleanups in header files: avoid nested includes.
 30. Fix posterLib LP64 issues.
 29. Implemented a real-time logging library for RTAI. Added logInit and
     logEnd. Bumped portLib minor.
 28. Updated RTAI/user-space to handle multi-threaded linux processes.
     It now uses one real-time task per thread.
 27. Added devfs support for RTAI.
 26. install-sh does not overwrite files if they were not modified.
 25. Preliminary RTAI/user-space support. posterLib requires a more
     complicated implementation: as of now, posters can be shared between
     RTAI-tasks and linux processes but this breaks the real-time
     properties of RTAI-tasks.
 24. Added taskDelete() regression test.
 23. Added --with-debug option to configure script. It allows the
     selection of independant components to debug.
 22. Added h2devClean() in comLib. This was previously defined in h2.c
     but it was problematic for user-space RTAI version. Bumped comLib

src/comLib/h2timerLib.c
/*
 * Copyright (c) 1990, 2003-2006 CNRS/LAAS
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/* DESCRIPTION :
   This is the library providing timers, used for task scheduling,
   with programmable time slices.
*/

#include "pocolibs-config.h"

#include <string.h>

#include "portLib.h"
#include "errnoLib.h"
#include "semLib.h"
#include "wdLib.h"
#include "h2timerLib.h"

static const H2_ERROR h2timerLibH2errMsgs[] = H2_TIMER_LIB_H2_ERR_MSGS;

/* #define COMLIB_DEBUG_H2TIMERLIB */

#ifdef COMLIB_DEBUG_H2TIMERLIB
# define LOGDBG(x)     logMsg x
#else
# define LOGDBG(x)
#endif

static int delayCount;			/* global synchronisation counter */
static SEM_ID timerMutex;
static WDOG_ID timerWd;			/* timer watchdog */
static H2TIMER timerTab[NMAX_TIMERS];	/* array of timers */
static BOOL h2timerInited = FALSE;

static void timerInt(int arg); /* timer handler procedure */

/******************************************************************************
*
*   h2timerInit	 -  Initializes the timer library
*
*   Description:
*   Initializes the timer library, starts the watchdog that genrate time
*   slices interrupts.
*
*   Returns: OK or ERROR
*/
STATUS
h2timerInit(void)
{
	int nTimer;

	LOGDBG(("comLib:h2timerLib:h2timerInit: ---beginning\n"));

	/* Create watchdog and mutex semaphore */
	timerMutex = semMCreate(0);
	semTake(timerMutex, WAIT_FOREVER);

	if ((timerWd = wdCreate ()) == NULL) {
		semGive(timerMutex);
		return ERROR;
	}

	/* Reset timer array */
	memset (timerTab, 0, sizeof(timerTab));

	/* allocate a synchronisation variable for each timer */
	for (nTimer = 0; nTimer < NMAX_TIMERS; nTimer++) {
		timerTab[nTimer].semSync = semBCreate(SEM_Q_PRIORITY,
		    SEM_EMPTY);
		if (timerTab[nTimer].semSync == NULL) {
			logMsg("h2timerInit: error semBCreate %d\n", nTimer);
			return ERROR;
		}
	}
	/* reset global delay counter */
	delayCount = 0;

	/* done with shared data */
	semGive(timerMutex);

	/* Start the watchdog */
	wdStart (timerWd, 1, (FUNCPTR) timerInt, 0);

	LOGDBG(("comLib:h2timerLib:h2timerInit: ---end\n"));


	/* record errors msg */
	h2recordErrMsgs("h2timerInit", "h2timerLib", M_h2timerLib,
	    sizeof(h2timerLibH2errMsgs)/sizeof(H2_ERROR),
	    h2timerLibH2errMsgs);

	h2timerInited = TRUE;
	return OK;
}


/******************************************************************************
*
*   h2timerEnd	-  Cleanup timers
*
*   Returns: OK or ERROR
*/
STATUS
h2timerEnd(void)
{
	int nTimer;

	if (h2timerInited != TRUE) return OK;

	/* Stop global watch dog */
	wdDelete (timerWd);

	semTake(timerMutex, WAIT_FOREVER);
	for (nTimer = 0; nTimer < NMAX_TIMERS; nTimer++) {
		semDelete(timerTab[nTimer].semSync);
		timerTab[nTimer].semSync = NULL;
	}
	semGive(timerMutex);

	/* Free global semaphore */
	semDelete(timerMutex);

	h2timerInited = FALSE;
	return OK;
}


/*****************************************************************************
*
*   h2timerAlloc  -  Allocate a timer
*
*   Description:
*   Allocate memory and initialize a timer structure
*
*   Returns: the new timer id or NULL
*/

H2TIMER_ID
h2timerAlloc(void)
{
	int nTimer;			/* index of the allocted timer */
	H2TIMER_ID timerId;		/* Timer identifier */

	if (h2timerInited != TRUE) {
		errnoSet(S_h2timerLib_TIMER_NOT_INIT);
		return NULL;
	}

	/* Start manipulating shared data */
	semTake(timerMutex, WAIT_FOREVER);

	timerId = timerTab;

	/* Find a free timer */
	for (nTimer = 0; timerId->flagInit == H2TIMER_FLG_INIT; nTimer++) {
		/* No free timer : error */
		if (nTimer == NMAX_TIMERS) {
			errnoSet (S_h2timerLib_TOO_MUCH_TIMERS);
			semGive(timerMutex);
			return (NULL);
		}
		timerId = (H2TIMER_ID) timerId + 1;
	} /* for */

	/* Fill the new structure */
	timerId->status = STOPPED_TIMER;
	timerId->flagInit = H2TIMER_FLG_INIT;

	/* Done with shared data */
	semGive(timerMutex);

	LOGDBG(("comLib:h2timerLib:h2timerAlloc: alloc %#x\n", timerId));
	return (timerId);
}

/******************************************************************************
*
*   h2timerStart  -  Start a timer
*
*   Description:
*   Starts a timer with a given delay
*   Warning: the period should be a multiple or a divisor of the
*	     maximum delay DELAY_MAX
*
*   Retourns: OK or ERROR
*/
STATUS
h2timerStart(H2TIMER_ID timerId,
    int period,
    int delay)
{
	/* Check if the timer is initialized */
	if (timerId->flagInit != H2TIMER_FLG_INIT) {
		errnoSet (S_h2timerLib_TIMER_NOT_INIT);
		return (ERROR);
	}

	/* You can only start a stopped timer */
	if (timerId->status != STOPPED_TIMER) {
		errnoSet (S_h2timerLib_NOT_STOPPED_TIMER);
		return (ERROR);
	}

	/* Check period's validity XXX */
	if ((period <= 0)) {
		logMsg ("h2timerStart: period (%d) must be positive\n", 
		    period);
		errnoSet (S_h2timerLib_BAD_PERIOD);
		return (ERROR);
	}

	/* Check delay */
	if (delay < 0 || delay >= MAX_DELAY) {
		errnoSet (S_h2timerLib_BAD_DELAY);
		return (ERROR);
	}

	/* Fill the timer structure */
	timerId->period = period;
	timerId->delay = delay;
	timerId->count = 0;
	timerId->status = WAIT_DELAY;

	LOGDBG(("h2timerStart %d %d\n", period, delay));
	return (OK);
}

/******************************************************************************
*
*   h2timerPause  -  Wait for timer expiration
*
*   Description:
*   Waits for timer expiration, given by freeing the synchronisation variable
*
*   Returns: OK or ERROR
*/
STATUS
h2timerPause(H2TIMER_ID timerId)
{
	/* Check for timer initialization */
	if (timerId->flagInit != H2TIMER_FLG_INIT) {
		errnoSet (S_h2timerLib_TIMER_NOT_INIT);
		return (ERROR);
	}

	/* Wait until the semaphore is released */
	LOGDBG(("h2timerPause: got lock\n"));
	semTake(timerId->semSync, WAIT_FOREVER);

	/*
	 * XXXX
	 * Flush the semaphore to make sure next call will block
	 */
	while (semTake(timerId->semSync, NO_WAIT) == OK)
		;
	return (OK);

}

/******************************************************************************
*
*   h2timerPauseReset  -  Wait timer expiration and reset it
*
*   Description:
*   Waits for the timer expiration and then set it's value to 0 to keep
*   synchronisation
*
*   Returns: OK or ERROR
*/
STATUS
h2timerPauseReset(H2TIMER_ID timerId)
{
	/*  Check timer initialisation */
	if (timerId->flagInit != H2TIMER_FLG_INIT) {
		errnoSet (S_h2timerLib_TIMER_NOT_INIT);
		return (ERROR);
	}

	/* Wait until the semaphore is release */
	semTake (timerId->semSync, WAIT_FOREVER);

	timerId->count = 0;
	return (OK);
}

/******************************************************************************
*
*   h2timerStop	 -  Stop a timer
*
*
*   Returns: OK or ERROR
*/

STATUS
h2timerStop(H2TIMER_ID timerId)
{
	/* Check timer initialization */
	if (timerId->flagInit != H2TIMER_FLG_INIT) {
		errnoSet (S_h2timerLib_TIMER_NOT_INIT);
		return (ERROR);
	}

	/* Move to stopped state */
	timerId->status = STOPPED_TIMER;
	return (OK);
}

/******************************************************************************
*
*   h2timerChangePeriod	 -  Change the period of a timer
*
*
*   Returns: OK or ERROR
*/
STATUS
h2timerChangePeriod(H2TIMER_ID timerId, int period)
{
	/* Check timer initialization */
	if (timerId->flagInit != H2TIMER_FLG_INIT) {
		errnoSet (S_h2timerLib_TIMER_NOT_INIT);
		return (ERROR);
	}

	/* Check that the timer is running */
	if (timerId->status == STOPPED_TIMER) {
		errnoSet (S_h2timerLib_STOPPED_TIMER);
		return (ERROR);
	}

	/* Change period */
	timerId->period = period;
	/* Adjust current count if needed */
	timerId->count = timerId->count % period;
	return (OK);
}

/******************************************************************************
*
*   h2timerFree	 -  Free a timer structure
*
*   Retourns: OK or ERROR
*/

STATUS
h2timerFree(H2TIMER_ID timerId)
{
	/* Check that the timer is initialized */
	if (timerId->flagInit != H2TIMER_FLG_INIT) {
		errnoSet (S_h2timerLib_TIMER_NOT_INIT);
		return (ERROR);
	}

	/* Stop the timer and mark it not initialized */
	timerId->status = STOPPED_TIMER;
	timerId->flagInit = FALSE;
	return (OK);
}

/*****************************************************************************
*
*   timerInt  -	 Handler of the watchdog interrupt
*
*   Description:
*   Check for expired timers, release their synchronisation semaphores
*
*   Return: Nothing
*/
static void
timerInt(int arg)
{
	int nTimer;		  /* timer index */
	H2TIMER_ID timerId;	  /* Timer array pointer */

	LOGDBG(("comLib:h2timerLib:timerInt: delayCount %d\n", delayCount));

	/* Bump delay counter */
	if (++delayCount >= MAX_DELAY)
		delayCount = 0;

	/* Restart the watchdog */
	wdStart (timerWd, 1, (FUNCPTR) timerInt, 0);

	/* Loop on all timers */
	timerId = timerTab;
	for (nTimer = 0; nTimer < NMAX_TIMERS; nTimer++) {
		/* Only check initialized timers */
		if (timerId->flagInit == H2TIMER_FLG_INIT) {
			/* Timer running */
			if (timerId->status == RUNNING_TIMER) {
				LOGDBG((" -- count(%d) = %d (period = %d)\n", nTimer,
					timerId->count, timerId->period));

				/* bump counter */
				if (++timerId->count >= timerId->period) {
					/* reset it */
					timerId->count = 0;
					LOGDBG((" -- timer %d triggered\n",
						nTimer));

					/* Release the semaphore */
					semGive (timerId->semSync);

				}
			} else if (timerId->status >= WAIT_DELAY) {

				/* Waiting initial delay */

				/* If delay is reached */
				if (delayCount == timerId->delay) {
					/* reset the counter */
					timerId->count = 0;

					/* Start running */
					LOGDBG((" -- timer %d starting\n",
						nTimer));
					timerId->status = RUNNING_TIMER;
					semGive (timerId->semSync);
				}
			}
		}

		/* next */
		timerId++;
	}
}
